### PR TITLE
Customize layout from settings

### LINKED ISSUE #8960: Jlab could provide a way to select predefined layouts where widgets are always docked together in a certain way
### Problem

The issue can be phrased like:

Is there a way of having some sort of predefined layouts so I can assign a given type of:
- Documents
- Widget
- Output
- Viewer
- Terminal

 to a fixed position in JLab?

Maybe I want all editors to be tabbed together on the left side, and all terminals tabbed in the bottom and outputs that result in some sort of plot (is this a mimerenderer?) always in the bottom right and the contextual help and csv viewer on the top right.

Having to open JLab and organize it in a certain way every time I open it, and hope that workspaces are going to do the right thing (this is often not the case) can be very frustrating and time-consuming.

Jlab

Why did the notebook I opened got tabbed with the terminal? Because it was the active/focused widget?

Having things in a given place is a time tested pattern that can be seen over and over in many IDEs like:

#### Matlab

image

#### RStudio

image

#### Spyder
image

### Eclipse

image

etc...

### Proposed Solution

Maybe I want to have something like this and call it the Rstudio layout (or something)

image

I could use a GridLayout so something like the image can be expressed as:

```typescript
rstudioLayout = GridLayout();
layout.addArea(row=0, column=0, name= “Top Left”, default=true,  [“notebook”, “editor”], default=true);
layout.addArea(row=0, column=1, name=“Top Right”,  [“contextual_help”]);
layout.addArea(row=1, column=0, name=“Bottom left”, [“console”]);
layout.addArea(row=1, column=1, name=“Bottom Right”, [“terminal”]);

layoutManager.register("RStudio", rstudioLayout);
```

This interface is inspired by https://doc.qt.io/qt-5/qgridlayout.html and could/would also allow to define column and row stretches to really provide a way to express any type of complex layout.

### PR DESCRIPTION
**Pull request by fcollonval**
## References

Fix #8960

demo_customize_layout

### CODE REVIEW

**Comment by fcollonval** [comment]
cc @jtpio @ellisonbg @bollwyvl here is a proposition to get widget layout customization. I think we could use such approach in RetroLab too to overcome supporting extension panel positionning in it. For example, in retrolab we could allow three state for sidebar: `left`, `right`, `treeView` (depending if the panel needs to be positioned on a notebook sidebar or in the initial tree view).

**Comment by bollwyvl** [comment]
fascinating, will take a look!

**Comment by bollwyvl** [comment]
One thought on this: at the end of it, I would hope there wouldn't be much "special" stuff that was not introspectable from outside that couldn't be resolved until there was a JS interpreter and hot DOM.

By that I mean: instead of the magic string `terminals` it might be something namespaced by `@jupyterlab/terminals`, as one of the more (accidentally) powerful patterns we have is that `@{org}/{pkg}:{plugin}` is configurable if and only if there exists a file called `$PREFIX/labextensions/${org}/${pkg}/{schemaDir}/{plugin}.json` (ignoring for the moment that you have to look up `schemaDir`). There, schema has _also_ given us predictable human-readable things to know, as well as lots of configuration. 

A less-discoverable language we already have is [lumino commands](https://github.com/jupyterlab/lumino/issues/58#issuecomment-963103017). This is also, semi-accidentally, surfaced by the [workspace](https://github.com/jupyterlab/jupyterlab/issues/6013). It suffers, however, from the commands _not_ being registered by a controlled namespace.

I imagine the final form of this at some point in the future being an abstract data shell that is a mapping of (the possibility of a) components to a set of regions, with a descriptive-enough language that it would be straightforward to build a visual tool that could discover, configure, and re-combine all available screen elements into:

- a dashboard
- a [single-cell](https://github.com/jupyterlite/jupyterlite/issues/148)
- a presentation
 
...where any of them could include self-modifying tools, such as a command palette (with the now-hidden group metadata restored, and expanded).

Practically, I imagine this would change `shell.add` to _not_ specify (or at least require/always respect) `area`, as the user-configured shell would decide, based on the widget type, where that-which-is-being-added should go, especially if the chosen `area` didn't exist. 

also cc @ajbozarth... this isn't exactly the launcher issue (never made the issue, sorry), but it's a related concern, as launchers would probably have to be aware of this kind of system.

**Comment by fcollonval** [comment]
please update documentation snapshots

**Comment by fcollonval** [comment]
please update documentation snapshots

**Comment by fcollonval** [comment]
CI failure tracked in https://github.com/jupyterlab/jupyterlab/pull/12552

**Comment by jtpio** [comment]
Nice, thanks @fcollonval for iterating on this!

Looks like there are a few conflicts to fix.

File: docs/source/user/interface_customization.rst [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+- TSV Viewer: *TSV Viewer* -> *toolbar*
+
+Those settings are accessible through the interactive *Settings Editor*. But you
+may find more comfortable to use the *Advanced Settings Editor* (accessible by clicking
+on *JSON Settings Editor* button in the top right corner of the *Settings Editor*).
+
+Your ``toolbar`` customization will be merged with the default menu bar definition (this
+is a different behavior as the other settings that are overriding the default settings).
+So to remove a default item, you will need to disable it. To do that, you will need
+to set the toolbar item ``disabled`` attribute to ``true``; a item is uniquely identified
```
Comment (trungleduc) [inline docs/source/user/interface_customization.rst:109]:
```suggestion
to set the toolbar item ``disabled`` attribute to ``true``; an item is uniquely identified
```

File: docs/source/user/interface_customization.rst [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+.. _menu-bar-customization:
+
+Menu Bar
+--------
+
+The top menu bar definition can be customized through the *Advanced Settings Editor*. You can
+access it by opening the *Settings Editor* from the *Settings* menu, then click on the
+*JSON Settings Editor* in the top right corner.
+
+In the *Advanced Settings Editor*, you will need to the select the *Main Menu* section in
```
Comment (trungleduc) [inline docs/source/user/interface_customization.rst:158]:
```suggestion
In the *Advanced Settings Editor*, you will need to select the *Main Menu* section in
```

File: docs/source/user/interface_customization.rst [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+        // Add new terminals in the down area in simple mode
+        "Terminal": { "area": "down" }
+      },
+      "multiple": {
+        // Add new terminals in the right sidebar in default mode
+        "Terminal": { "area": "right" }
+      }
+    }
+
+The above example, will results in the following changes:
```
Comment (trungleduc) [inline docs/source/user/interface_customization.rst:74]:
```suggestion
The above example will result in the following changes:
```

File: packages/application-extension/src/index.tsx [hunk 4]
```diff
@@ -1356,50 +1405,13 @@ namespace Private {
...
-        labShell.layoutModified.connect(update);
-        update(labShell, initial);
-      }
-    );
-
     // Add a command to switch a side panels's side
     app.commands.addCommand(CommandIDs.switchSidebar, {
       label: trans.__('Switch Sidebar Side'),
-      execute: () => {
+      execute: async () => {
```
Comment (trungleduc) [inline packages/application-extension/src/index.tsx:1414]: Do we not need `async` here?

**Review by trungleduc — COMMENTED** [review]
Thanks @fcollonval, minor suggestions on the doc then it looks great to me!

File: packages/application-extension/src/index.tsx (hunk 4 above)
Comment (fcollonval) [reply packages/application-extension/src/index.tsx:1414]: Good catch

**Comment by fcollonval** [comment]
Thanks @trungleduc for the review

**Comment by jtpio** [comment]
Nice this is looking good :+1: 

Just tested on Binder and it looks like the terminal is still opened in main area when created from the launcher:

https://user-images.githubusercontent.com/591645/169058846-bbd9e4cb-158c-4cfb-b40a-e474babd0d29.mp4

**Comment by jtpio** [comment]
Nice thanks!

Apart from that it's looking good. I think we can get this in so it can be released in the next 4.0 pre-release and tested in Notebook v7.

**Comment by jtpio** [comment]
Check again on Binder and looking good:

https://user-images.githubusercontent.com/591645/169076381-eec06930-2e26-408a-b0b9-e3fcda03042e.mp4

**Review by jtpio — APPROVED** [review]
Thanks!

**Comment by fcollonval** [comment]
Thanks for the reviews

**Pull request merged**